### PR TITLE
KAFKA-5933: Move timeout and validate_only protocol fields into CommonFields class

### PR DESCRIPTION
**Pull request by scholzj**
Most of the fields which are shared by multiple protocol messages (requests ad responses) are in the CommonFields class. However there are still some fields used multiple times which are not there yet:
* `timeout`
* `validate_only`

It would be good to move also these two fields into the CommonFields class so that they can be easily shared by different messages.

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/requests/CreateTopicsRequest.java [hunk 1]
```diff
@@ -249,9 +245,9 @@ public CreateTopicsRequest(Struct struct, short version) {
         }
 
         this.topics = topics;
-        this.timeout = struct.getInt(TIMEOUT_KEY_NAME);
-        if (struct.hasField(VALIDATE_ONLY_KEY_NAME))
-            this.validateOnly = struct.getBoolean(VALIDATE_ONLY_KEY_NAME);
+        this.timeout = struct.get(TIMEOUT);
+        if (struct.hasField(VALIDATE_ONLY))
+            this.validateOnly = struct.get(VALIDATE_ONLY);
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/requests/CreateTopicsRequest.java]: If we add a `getOrElse`, this becomes simpler.

Comment (scholzj) [reply clients/src/main/java/org/apache/kafka/common/requests/CreateTopicsRequest.java]: Fixed.

**Comment by scholzj** [comment]
@ijuma Do you have some more comments on this?

File: clients/src/main/java/org/apache/kafka/common/protocol/types/Struct.java [hunk 2]
```diff
@@ -123,6 +127,12 @@ public String getOrElse(Field.Str field, String alternative) {
         return alternative;
     }
 
+    public Boolean getOrElse(Field.Boolean field, Boolean alternative) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/protocol/types/Struct.java]: Seems like `alternative` should be `boolean`?

File: clients/src/main/java/org/apache/kafka/common/protocol/types/Struct.java [hunk 3]
```diff
@@ -288,6 +298,10 @@ public Struct set(Field.Int16 def, short value) {
         return set(def.name, value);
     }
 
+    public Struct set(Field.Boolean def, Boolean value) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/protocol/types/Struct.java]: Seems like `value` should be `boolean`?

File: clients/src/main/java/org/apache/kafka/common/protocol/types/Struct.java [hunk 4]
```diff
@@ -99,6 +99,10 @@ public String get(Field.NullableStr field) {
         return getString(field.name);
     }
 
+    public Boolean get(Field.Boolean field) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/protocol/types/Struct.java]: Should the result type be `boolean`? @hachikuji, was it intentional that the result type of various `get` methods is returning the boxed type?

**Review by ijuma — COMMENTED** [review]
Sorry for the delay, left a few minor comments.

File: clients/src/main/java/org/apache/kafka/common/protocol/types/Struct.java (hunk 4 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/common/protocol/types/Struct.java]: The value is already boxed, but I can't think of a great reason why we should return the boxed type. Other than strings, the primitive types don't allow null.

Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/protocol/types/Struct.java]: That's what I was thinking, it's more obvious that null is not expected if it's a primitive.

Comment (scholzj) [reply clients/src/main/java/org/apache/kafka/common/protocol/types/Struct.java]: @ijuma @hachikuji Ok, I changed it to `boolean`. However, The other methods seem to be using Long, Integer, Short etc. as well. Do we want to change them with this PR as well? Or raise a JIRA for it?

Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/protocol/types/Struct.java]: Let's just do it in this PR. Thanks!

Comment (scholzj) [reply clients/src/main/java/org/apache/kafka/common/protocol/types/Struct.java]: Ok, I changed it also for Long, Short, Integer and Byte. It should be fixed now.

File: clients/src/main/java/org/apache/kafka/common/protocol/CommonFields.java [hunk 5]
```diff
@@ -26,6 +26,11 @@
     public static final Field.Int32 PARTITION_ID = new Field.Int32("partition", "Topic partition id");
     public static final Field.Int16 ERROR_CODE = new Field.Int16("error_code", "Response error code");
     public static final Field.NullableStr ERROR_MESSAGE = new Field.NullableStr("error_message", "Response error message");
+    public static final Field.Boolean VALIDATE_ONLY = new Field.Boolean("validate_only",
+            "If this is true, the request will be validated, but the " +
+                     "topic won't be created.");
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/protocol/CommonFields.java]: Seems like "topic won't be created" is specific to one of the requests.

Comment (scholzj) [reply clients/src/main/java/org/apache/kafka/common/protocol/CommonFields.java]: Fixed.

**Comment by scholzj** [comment]
Hi @ijuma ... do you have some more comments for this PR?

**Review by ijuma — COMMENTED** [review]
Sorry for the delay. One concern is that we're losing request specific information on what the timeout means. For example:

```java
"The time in ms to wait for a topic to be completely created on the controller node"
```

@hachikuji any thoughts on that?

**Comment by hachikuji** [comment]
@ijuma Yeah, this is where the `CommonFields` pattern sort of breaks down. Perhaps we need a facility for the schema to override the doc string?

**Pull request closed without merging**